File: routa-core/src/main/kotlin/com/phodal/routa/core/mcp/RoutaMcpToolManager.kt
package com.phodal.routa.core.mcp

import com.phodal.routa.core.model.AgentRole
import com.phodal.routa.core.model.CompletionReport
import com.phodal.routa.core.model.ModelTier
import com.phodal.routa.core.tool.AgentTools
import io.modelcontextprotocol.kotlin.sdk.server.Server
import io.modelcontextprotocol.kotlin.sdk.types.*
import kotlinx.serialization.json.*

/**
 * Registers the 10 Routa coordination tools with an MCP [Server].
 *
 * This exposes the coordination tools over the Model Context Protocol,
 * allowing any MCP-compatible client (e.g., Claude, Cursor, VS Code)
 * to use them for multi-agent coordination.
 *
 * Usage:
 * ```kotlin
 * val mcpServer = Server(...)
 * RoutaMcpToolManager(agentTools, "my-workspace").registerTools(mcpServer)
 * ```
 */
class RoutaMcpToolManager(
    private val agentTools: AgentTools,
    private val defaultWorkspaceId: String,
) {

    /**
     * Register all 10 coordination tools with the MCP server.
     */
    fun registerTools(server: Server) {
        // Core coordination tools
        registerListAgents(server)
        registerReadAgentConversation(server)
        registerCreateAgent(server)
        registerDelegateTask(server)
        registerMessageAgent(server)
        registerReportToParent(server)
        // Task-agent lifecycle tools
        registerWakeOrCreateTaskAgent(server)
        registerSendMessageToTaskAgent(server)
        registerGetAgentStatus(server)
        registerGetAgentSummary(server)
    }

    private fun registerListAgents(server: Server) {
        server.addTool(
            name = "list_agents",
            description = "List all agents in the workspace. Shows each agent's ID, name, role " +
                "(ROUTA/CRAFTER/GATE), status, and parent.",
            inputSchema = ToolSchema(
                properties = buildJsonObject {
                    putJsonObject("workspaceId") {
                        put("type", "string")
                        put("description", "The workspace ID (uses default if empty)")
                    }
                },
                required = emptyList()
            )
        ) { request ->
            val workspaceId = request.arguments?.get("workspaceId")
                ?.jsonPrimitive?.contentOrNull ?: defaultWorkspaceId
            val result = agentTools.listAgents(workspaceId)
            toCallToolResult(result)
        }
    }

    private fun registerReadAgentConversation(server: Server) {
        server.addTool(
            name = "read_agent_conversation",
            description = "Read another agent's conversation history. Supports full history, " +
                "last N messages, or a specific turn range. Use to review work or avoid conflicts.",
            inputSchema = ToolSchema(
                properties = buildJsonObject {
                    putJsonObject("agentId") {
                        put("type", "string")
                        put("description", "ID of the agent whose conversation to read")
                    }
                    putJsonObject("lastN") {
                        put("type", "integer")
                        put("description", "Only return the last N messages (optional)")
                    }
                    putJsonObject("includeToolCalls") {
                        put("type", "boolean")
                        put("description", "Include tool calls in the output (default: true)")
                    }
                },
                required = listOf("agentId")
            )
        ) { request ->
            val args = request.arguments ?: JsonObject(emptyMap())
            val agentId = args["agentId"]!!.jsonPrimitive.content
            val lastN = args["lastN"]?.jsonPrimitive?.intOrNull
            val includeToolCalls = args["includeToolCalls"]?.jsonPrimitive?.booleanOrNull ?: true

            val result = agentTools.readAgentConversation(
                agentId = agentId,
                lastN = lastN,
                includeToolCalls = includeToolCalls,
            )
            toCallToolResult(result)
        }
    }

    private fun registerCreateAgent(server: Server) {
        server.addTool(
            name = "create_agent",
            description = "Create a new agent. Roles: ROUTA (coordinator), CRAFTER (implementor), " +
                "GATE (verifier). Model tiers: SMART (planning/verification), FAST (implementation).",
            inputSchema = ToolSchema(
                properties = buildJsonObject {
                    putJsonObject("name") {
                        put("type", "string")
                        put("description", "Human-readable name for the agent")
                    }
                    putJsonObject("role") {
                        put("type", "string")
                        put("description", "Agent role: ROUTA, CRAFTER, or GATE")
                        putJsonArray("enum") {
                            add("ROUTA"); add("CRAFTER"); add("GATE")
                        }
                    }
                    putJsonObject("parentId") {
                        put("type", "string")
                        put("description", "ID of the parent agent (optional)")
                    }
                    putJsonObject("modelTier") {
                        put("type", "string")
                        put("description", "Model tier: SMART or FAST (optional, uses role default)")
                        putJsonArray("enum") {
                            add("SMART"); add("FAST")
                        }
                    }
                },
                required = listOf("name", "role")
            )
        ) { request ->
            val args = request.arguments ?: JsonObject(emptyMap())
            val name = args["name"]!!.jsonPrimitive.content
            val roleStr = args["role"]!!.jsonPrimitive.content
            val parentId = args["parentId"]?.jsonPrimitive?.contentOrNull
            val modelTierStr = args["modelTier"]?.jsonPrimitive?.contentOrNull

            val role = try {
                AgentRole.valueOf(roleStr.uppercase())
            } catch (e: IllegalArgumentException) {
                return@addTool CallToolResult(
                    content = listOf(TextContent(text = "Invalid role: $roleStr")),
                    isError = true
                )
            }

            val modelTier = modelTierStr?.let {
                try { ModelTier.valueOf(it.uppercase()) } catch (e: IllegalArgumentException) { null }
            }

            val result = agentTools.createAgent(
                name = name,
                role = role,
                workspaceId = defaultWorkspaceId,
                parentId = parentId,
                modelTier = modelTier,
            )
            toCallToolResult(result)
        }
    }

    private fun registerDelegateTask(server: Server) {
        server.addTool(
            name = "delegate_task",
            description = "Delegate a task to a specific agent. Assigns the task and activates the agent.",
            inputSchema = ToolSchema(
                properties = buildJsonObject {
                    putJsonObject("agentId") {
                        put("type", "string")
                        put("description", "ID of the agent to delegate to")
                    }
                    putJsonObject("taskId") {
                        put("type", "string")
                        put("description", "ID of the task to assign")
                    }
                    putJsonObject("callerAgentId") {
                        put("type", "string")
                        put("description", "ID of the agent performing the delegation")
                    }
                },
                required = listOf("agentId", "taskId", "callerAgentId")
            )
        ) { request ->
            val args = request.arguments ?: JsonObject(emptyMap())
            val result = agentTools.delegate(
                agentId = args["agentId"]!!.jsonPrimitive.content,
                taskId = args["taskId"]!!.jsonPrimitive.content,
                callerAgentId = args["callerAgentId"]!!.jsonPrimitive.content,
            )
            toCallToolResult(result)
        }
    }

    private fun registerMessageAgent(server: Server) {
        server.addTool(
            name = "send_message_to_agent",
            description = "Send a message to another agent for inter-agent communication: " +
                "conflict reports, fix requests, additional context.",
            inputSchema = ToolSchema(
                properties = buildJsonObject {
                    putJsonObject("fromAgentId") {
                        put("type", "string")
                        put("description", "ID of the sending agent")
                    }
                    putJsonObject("toAgentId") {
                        put("type", "string")
                        put("description", "ID of the receiving agent")
                    }
                    putJsonObject("message") {
                        put("type", "string")
                        put("description", "The message content")
                    }
                },
                required = listOf("fromAgentId", "toAgentId", "message")
            )
        ) { request ->
            val args = request.arguments ?: JsonObject(emptyMap())
            val result = agentTools.messageAgent(
                fromAgentId = args["fromAgentId"]!!.jsonPrimitive.content,
                toAgentId = args["toAgentId"]!!.jsonPrimitive.content,
                message = args["message"]!!.jsonPrimitive.content,
            )
            toCallToolResult(result)
        }
    }

    private fun registerReportToParent(server: Server) {
        server.addTool(
            name = "report_to_parent",
            description = "Send a completion report to the parent agent. REQUIRED for all delegated " +
                "agents (Crafter and Gate). Include: what you did, verification results, risks.",
            inputSchema = ToolSchema(
                properties = buildJsonObject {
                    putJsonObject("agentId") {
                        put("type", "string")
                        put("description", "ID of the reporting agent")
                    }
                    putJsonObject("taskId") {
                        put("type", "string")
                        put("description", "ID of the task being reported on")
                    }
                    putJsonObject("summary") {
                        put("type", "string")
                        put("description", "1-3 sentence summary: what you did, verification, risks")
                    }
                    putJsonObject("filesModified") {
                        put("type", "array")
                        putJsonObject("items") { put("type", "string") }
                        put("description", "List of files that were modified")
                    }
                    putJsonObject("success") {
                        put("type", "boolean")
                        put("description", "Whether the task was completed successfully")
                    }
                },
                required = listOf("agentId", "taskId", "summary")
            )
        ) { request ->
            val args = request.arguments ?: JsonObject(emptyMap())
            val agentId = args["agentId"]!!.jsonPrimitive.content
            val filesModified = args["filesModified"]?.jsonArray
                ?.map { it.jsonPrimitive.content } ?: emptyList()
            val success = args["success"]?.jsonPrimitive?.booleanOrNull ?: true

            val report = CompletionReport(
                agentId = agentId,
                taskId = args["taskId"]!!.jsonPrimitive.content,
                summary = args["summary"]!!.jsonPrimitive.content,
                filesModified = filesModified,
                success = success,
            )

            // Log the report for verification
            println("🎯 MCP Tool Called: report_to_parent")
            println("   Agent: $agentId")
            println("   Task: ${report.taskId}")
            println("   Success: ${report.success}")
            println("   Summary: ${report.summary}")
            if (filesModified.isNotEmpty()) {
                println("   Files: ${filesModified.joinToString(", ")}")
            }

            val result = agentTools.reportToParent(agentId, report)
            toCallToolResult(result)
        }
    }

    private fun registerWakeOrCreateTaskAgent(server: Server) {
        server.addTool(
            name = "wake_or_create_task_agent",
            description = "Wake an existing agent or create a new one for a task. " +
                "Checks if the task has an active/pending agent and wakes it with the context message. " +
                "If no viable agent, creates a new Crafter and assigns it. " +
                "Use when task dependencies become ready.",
            inputSchema = ToolSchema(
                properties = buildJsonObject {
                    putJsonObject("taskId") {
                        put("type", "string")
                        put("description", "ID of the task to wake or create an agent for")
                    }
                    putJsonObject("contextMessage") {
                        put("type", "string")
                        put("description", "Message with synthesized context from completed dependencies")
                    }
                    putJsonObject("callerAgentId") {
                        put("type", "string")
                        put("description", "ID of the calling agent (for auditing)")
                    }
                    putJsonObject("agentName") {
                        put("type", "string")
                        put("description", "Optional custom name for a new agent")
                    }
                    putJsonObject("modelTier") {
                        put("type", "string")
                        put("description", "Model tier for a new agent: SMART or FAST (optional)")
                        putJsonArray("enum") {
                            add("SMART"); add("FAST")
                        }
                    }
                },
                required = listOf("taskId", "contextMessage", "callerAgentId")
            )
        ) { request ->
            val args = request.arguments ?: JsonObject(emptyMap())
            val modelTierStr = args["modelTier"]?.jsonPrimitive?.contentOrNull
            val modelTier = modelTierStr?.let {
                try {
                    com.phodal.routa.core.model.ModelTier.valueOf(it.uppercase())
                } catch (e: IllegalArgumentException) { null }
            }

            val result = agentTools.wakeOrCreateTaskAgent(
                taskId = args["taskId"]!!.jsonPrimitive.content,
                contextMessage = args["contextMessage"]!!.jsonPrimitive.content,
                callerAgentId = args["callerAgentId"]!!.jsonPrimitive.content,
                workspaceId = defaultWorkspaceId,
                agentName = args["agentName"]?.jsonPrimitive?.contentOrNull,
                modelTier = modelTier,
            )
            toCallToolResult(result)
        }
    }

    private fun registerSendMessageToTaskAgent(server: Server) {
        server.addTool(
            name = "send_message_to_task_agent",
            description = "Send a message to the agent working on a specific task. " +
                "More convenient than send_message_to_agent — you only need the task ID. " +
                "Use to ask corrections, provide context, or request changes.",
            inputSchema = ToolSchema(
                properties = buildJsonObject {
                    putJsonObject("taskId") {
                        put("type", "string")
                        put("description", "ID of the task whose agent should receive the message")
                    }
                    putJsonObject("message") {
                        put("type", "string")
                        put("description", "The message content. Be specific about what changes or corrections you need.")
                    }
                    putJsonObject("callerAgentId") {
                        put("type", "string")
                        put("description", "ID of the sending agent")
                    }
                },
                required = listOf("taskId", "message", "callerAgentId")
            )
        ) { request ->
            val args = request.arguments ?: JsonObject(emptyMap())
            val result = agentTools.sendMessageToTaskAgent(
                taskId = args["taskId"]!!.jsonPrimitive.content,
                message = args["message"]!!.jsonPrimitive.content,
                callerAgentId = args["callerAgentId"]!!.jsonPrimitive.content,
            )
            toCallToolResult(result)
        }
    }

    private fun registerGetAgentStatus(server: Server) {
        server.addTool(
            name = "get_agent_status",
            description = "Get detailed status of a specific agent including current status, " +
                "role, message count, assigned tasks, and timestamps.",
            inputSchema = ToolSchema(
                properties = buildJsonObject {
                    putJsonObject("agentId") {
                        put("type", "string")
                        put("description", "ID of the agent to check")
                    }
                },
                required = listOf("agentId")
            )
        ) { request ->
            val args = request.arguments ?: JsonObject(emptyMap())
            val result = agentTools.getAgentStatus(
                agentId = args["agentId"]!!.jsonPrimitive.content,
            )
            toCallToolResult(result)
        }
    }

    private fun registerGetAgentSummary(server: Server) {
        server.addTool(
            name = "get_agent_summary",
            description = "Get a summary of what an agent did. Returns status, last response, " +
                "tool call counts, and assigned tasks. Quick overview before reading full conversation.",
            inputSchema = ToolSchema(
                properties = buildJsonObject {
                    putJsonObject("agentId") {
                        put("type", "string")
                        put("description", "ID of the agent to summarize")
                    }
                },
                required = listOf("agentId")
            )
        ) { request ->
            val args = request.arguments ?: JsonObject(emptyMap())
            val result = agentTools.getAgentSummary(
                agentId = args["agentId"]!!.jsonPrimitive.content,
            )
            toCallToolResult(result)
        }
    }

    private fun toCallToolResult(result: com.phodal.routa.core.tool.ToolResult): CallToolResult {
        return CallToolResult(
            content = listOf(TextContent(text = if (result.success) result.data else (result.error ?: "Unknown error"))),
            isError = !result.success
        )
    }
}

File: routa-js/src/core/mcp/routa-mcp-tool-manager.ts
/**
 * RoutaMcpToolManager - port of routa-core RoutaMcpToolManager.kt
 *
 * Registers all 12 AgentTools as MCP tools on an McpServer instance.
 * Each tool maps directly to an AgentTools method.
 */

import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z } from "zod";
import { AgentTools } from "../tools/agent-tools";
import { ToolResult } from "../tools/tool-result";

export class RoutaMcpToolManager {
  constructor(
    private tools: AgentTools,
    private workspaceId: string
  ) {}

  /**
   * Register all 12 coordination tools with the MCP server.
   */
  registerTools(server: McpServer): void {
    this.registerListAgents(server);
    this.registerReadAgentConversation(server);
    this.registerCreateAgent(server);
    this.registerDelegateTask(server);
    this.registerSendMessageToAgent(server);
    this.registerReportToParent(server);
    this.registerWakeOrCreateTaskAgent(server);
    this.registerSendMessageToTaskAgent(server);
    this.registerGetAgentStatus(server);
    this.registerGetAgentSummary(server);
    this.registerSubscribeToEvents(server);
    this.registerUnsubscribeFromEvents(server);
  }

  private registerListAgents(server: McpServer) {
    server.tool(
      "list_agents",
      "List all agents in the current workspace with their id, name, role, status, and parentId",
      {
        workspaceId: z.string().optional().describe("Workspace ID (uses default if omitted)"),
      },
      async (params) => {
        const result = await this.tools.listAgents(params.workspaceId ?? this.workspaceId);
        return this.toMcpResult(result);
      }
    );
  }

  private registerReadAgentConversation(server: McpServer) {
    server.tool(
      "read_agent_conversation",
      "Read conversation history of another agent. Use lastN for recent messages or startTurn/endTurn for a range.",
      {
        agentId: z.string().describe("ID of the agent whose conversation to read"),
        lastN: z.number().optional().describe("Number of recent messages to retrieve"),
        startTurn: z.number().optional().describe("Start turn number (inclusive)"),
        endTurn: z.number().optional().describe("End turn number (inclusive)"),
        includeToolCalls: z.boolean().optional().describe("Include tool call messages (default: false)"),
      },
      async (params) => {
        const result = await this.tools.readAgentConversation(params);
        return this.toMcpResult(result);
      }
    );
  }

  private registerCreateAgent(server: McpServer) {
    server.tool(
      "create_agent",
      "Create a new agent with a role (ROUTA=coordinator, CRAFTER=implementor, GATE=verifier)",
      {
        name: z.string().describe("Name for the new agent"),
        role: z.enum(["ROUTA", "CRAFTER", "GATE"]).describe("Agent role"),
        workspaceId: z.string().optional().describe("Workspace ID (uses default if omitted)"),
        parentId: z.string().optional().describe("Parent agent ID"),
        modelTier: z.enum(["SMART", "FAST"]).optional().describe("Model tier (default: SMART)"),
      },
      async (params) => {
        const result = await this.tools.createAgent({
          ...params,
          workspaceId: params.workspaceId ?? this.workspaceId,
        });
        return this.toMcpResult(result);
      }
    );
  }

  private registerDelegateTask(server: McpServer) {
    server.tool(
      "delegate_task",
      "Assign a task to an agent and activate it. The agent will begin working on the task.",
      {
        agentId: z.string().describe("ID of the agent to delegate to"),
        taskId: z.string().describe("ID of the task to delegate"),
        callerAgentId: z.string().describe("ID of the calling agent"),
      },
      async (params) => {
        const result = await this.tools.delegate(params);
        return this.toMcpResult(result);
      }
    );
  }

  private registerSendMessageToAgent(server: McpServer) {
    server.tool(
      "send_message_to_agent",
      "Send a message from one agent to another. The message is added to the target agent's conversation.",
      {
        fromAgentId: z.string().describe("ID of the sending agent"),
        toAgentId: z.string().describe("ID of the receiving agent"),
        message: z.string().describe("Message content"),
      },
      async (params) => {
        const result = await this.tools.messageAgent(params);
        return this.toMcpResult(result);
      }
    );
  }

  private registerReportToParent(server: McpServer) {
    server.tool(
      "report_to_parent",
      "Submit a completion report to the parent agent. Updates task status and notifies the parent.",
      {
        agentId: z.string().describe("ID of the reporting agent"),
        taskId: z.string().describe("ID of the completed task"),
        summary: z.string().describe("Summary of what was accomplished"),
        filesModified: z.array(z.string()).optional().describe("List of modified files"),
        verificationResults: z.string().optional().describe("Verification output"),
        success: z.boolean().describe("Whether the task was completed successfully"),
      },
      async (params) => {
        const result = await this.tools.reportToParent({
          agentId: params.agentId,
          report: {
            agentId: params.agentId,
            taskId: params.taskId,
            summary: params.summary,
            filesModified: params.filesModified,
            verificationResults: params.verificationResults,
            success: params.success,
          },
        });
        return this.toMcpResult(result);
      }
    );
  }

  private registerWakeOrCreateTaskAgent(server: McpServer) {
    server.tool(
      "wake_or_create_task_agent",
      "Wake an existing agent assigned to a task, or create a new Crafter agent if none exists.",
      {
        taskId: z.string().describe("ID of the task"),
        contextMessage: z.string().describe("Context message for the agent"),
        callerAgentId: z.string().describe("ID of the calling agent"),
        workspaceId: z.string().optional().describe("Workspace ID (uses default if omitted)"),
        agentName: z.string().optional().describe("Name for new agent (if created)"),
        modelTier: z.enum(["SMART", "FAST"]).optional().describe("Model tier for new agent"),
      },
      async (params) => {
        const result = await this.tools.wakeOrCreateTaskAgent({
          ...params,
          workspaceId: params.workspaceId ?? this.workspaceId,
        });
        return this.toMcpResult(result);
      }
    );
  }

  private registerSendMessageToTaskAgent(server: McpServer) {
    server.tool(
      "send_message_to_task_agent",
      "Send a message to the agent currently assigned to a task.",
      {
        taskId: z.string().describe("ID of the task"),
        message: z.string().describe("Message content"),
        callerAgentId: z.string().describe("ID of the calling agent"),
      },
      async (params) => {
        const result = await this.tools.sendMessageToTaskAgent(params);
        return this.toMcpResult(result);
      }
    );
  }

  private registerGetAgentStatus(server: McpServer) {
    server.tool(
      "get_agent_status",
      "Get the current status, message count, and assigned tasks for an agent.",
      {
        agentId: z.string().describe("ID of the agent"),
      },
      async (params) => {
        const result = await this.tools.getAgentStatus(params.agentId);
        return this.toMcpResult(result);
      }
    );
  }

  private registerGetAgentSummary(server: McpServer) {
    server.tool(
      "get_agent_summary",
      "Get a summary of an agent including last response, tool call counts, and active tasks.",
      {
        agentId: z.string().describe("ID of the agent"),
      },
      async (params) => {
        const result = await this.tools.getAgentSummary(params.agentId);
        return this.toMcpResult(result);
      }
    );
  }

  private registerSubscribeToEvents(server: McpServer) {
    server.tool(
      "subscribe_to_events",
      "Subscribe an agent to workspace events (AGENT_CREATED, TASK_COMPLETED, etc.)",
      {
        agentId: z.string().describe("ID of the subscribing agent"),
        agentName: z.string().describe("Name of the subscribing agent"),
        eventTypes: z.array(z.string()).describe("Event types to subscribe to"),
        excludeSelf: z.boolean().optional().describe("Exclude self-generated events (default: true)"),
      },
      async (params) => {
        const result = await this.tools.subscribeToEvents(params);
        return this.toMcpResult(result);
      }
    );
  }

  private registerUnsubscribeFromEvents(server: McpServer) {
    server.tool(
      "unsubscribe_from_events",
      "Remove an event subscription.",
      {
        subscriptionId: z.string().describe("ID of the subscription to remove"),
      },
      async (params) => {
        const result = await this.tools.unsubscribeFromEvents(params.subscriptionId);
        return this.toMcpResult(result);
      }
    );
  }

  // ─── Helpers ─────────────────────────────────────────────────────────

  private toMcpResult(result: ToolResult) {
    return {
      content: [
        {
          type: "text" as const,
          text: JSON.stringify(result.success ? result.data : { error: result.error }, null, 2),
        },
      ],
      isError: !result.success,
    };
  }
}
